tracing: Consolidate tracing into a new katatrace package

bporter816 opened the pull request:
Removes custom trace() functions defined across the project and creates
a single Trace() function in a new package called katatrace. Moves span
tag configuration for static tags to global variables in the package.

Fixes #3093 

Signed-off-by: Benjamin Porter <bporter816@gmail.com>

bporter816 commented:
@cmaf This is ready for review. It should need the `no-backport-needed` and `forward-port-needed` labels.

Something to point out is that this is a best effort at preserving the existing tags and loggers that were being used to do tracing and report errors. There are certainly some differences (some of the old `trace()` functions didn't report nil context issues at all, for example). We've made some structural changes, mainly moving `Trace()` and tracing.go into a new subpackage of katautils called katatrace. This was done to avoid a cyclical dependency between katautils and virtcontainers. katatracing doesn't import any kata packages.

One bigger change we made was to pass the logger from the caller into `Trace()`, to keep the fields specific to the caller. If that level of detail isn't needed, this could be removed.

We didn't make any changes to the `tracing-test.sh` testing script in the `tests` repo, but it still gave the same output as before our changes. Is there anything that needs to be done there?

egernst commented on virtcontainers/pkg/cloud-hypervisor/client/go.mod:
```diff
@@ -3,5 +3,7 @@ module github.com/GIT_USER_ID/GIT_REPO_ID
-	
+
 )
+
```
Why this change?

bporter816 replied:
Not sure about that one...I think it was an automatic editor change. I can revert the `go 1.13` line.

egernst commented on pkg/katautils/katatrace/tracing.go:
```diff
@@ -90,19 +103,32 @@ func StopTracing(ctx context.Context) {
 // Trace creates a new tracing span based on the specified name and parent
 // context.
-func Trace(parent context.Context, name string) (opentracing.Span, context.Context) {
+func Trace(parent context.Context, logger *logrus.Entry, name string, tags ...string) (opentracing.Span, context.Context) {
```
May be worth adding a note in the header re: tags. ie, its a slice of strings, but really its full of key-value pairs, right? Let's just make sure the caller is aware that there should be an even number of entries

bporter816 replied:
Fixed

egernst commented on virtcontainers/api.go:
```diff
@@ -20,7 +21,6 @@ import (
 	"github.com/kata-containers/runtime/virtcontainers/store"
 	"github.com/kata-containers/runtime/virtcontainers/types"
 	specs "github.com/opencontainers/runtime-spec/specs-go"
-	opentracing "github.com/opentracing/opentracing-go"
```
As we look @ potentially migrating to OpenTelemetry in the future, I appreciate being able to abstract this out of the rest of the code base...

egernst reviewed (commented):
I'm still working my way through it, but have a couple small bits of feedback in the meantime.

Ideally we'd split this to two commits: one for adding the (new) subpkg, one for updating the rest of the codebase to utilize it (though even there, you could split vc/cli, just for easy quick review). Not a major deal, maybe a nit,, but my preference.

@jodh-intel  can you PTAL too?

cmaf commented:
@bporter816 could you change the commit to match [this format](https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format)? Also a nit but if you could change the subsystem from logging to tracing, that would be helpful.

cmaf commented on pkg/katautils/config.go:
```diff
@@ -1218,7 +1218,7 @@ func LoadConfiguration(configPath string, ignoreLogging, builtIn bool) (resolved
 
 	config.Trace = tomlConf.Runtime.Tracing
-	tracing = config.Trace
+	Tracing = config.Trace
```
If we set the `tracingSet` variable in pkg/katautils/katatrace/tracing.go here, we can remove `tracing bool` from the `CreateTracer()` parameters.

bporter816 replied:
Fixed

cmaf commented on pkg/katautils/katatrace/tracing.go:
```diff
@@ -3,34 +3,47 @@
 
 // CreateTracer create a tracer
-func CreateTracer(name string) (opentracing.Tracer, error) {
+func CreateTracer(name string, tracing bool) (opentracing.Tracer, error) {
```
See above comment; we can remove `tracing bool` if tracingSet is assigned in [pkg/katautils/config.go](https://github.com/kata-containers/runtime/blob/master/pkg/katautils/config.go#L1220-L1221)

bporter816 replied:
Fixed

cmaf reviewed (changes requested):
Thanks @bporter816, a few comments about `CreateTracer()` and the `tracingSet` var

bporter816 commented:
@egernst @cmaf I addressed all the code-specific changes. If that looks good I'll squash and edit the commit format based on those comments.

egernst commented:
Looks fine to me now @bporter816

cmaf commented:
/test

cmaf reviewed (approved):
thanks @bporter816

cmaf commented:
@bporter816 one request, could you format the commits as described in our [docs](https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format) and make sure to add a description of the commit? Also, please split it into two commits as @egernst suggested.

bporter816 commented:
@cmaf @egernst I rebased and split the commits. There's a small conflict - not sure what your preference is for resolving those is so I'll leave it as is.

cmaf commented:
Hi @bporter816, could you please resolve the conflicts and force push the branch again?

bporter816 commented:
@cmaf Done, I think it should be good now.

cmaf commented:
/test

cmaf commented:
@egernst @jodh-intel would either of you be able to review?

@bporter816 do you think you will be able to port this to https://github.com/kata-containers/kata-containers? There was a PR that migrates to opentelemetry ([1352](https://github.com/kata-containers/kata-containers/pull/1352)) and another fixing span ordering ([1389](https://github.com/kata-containers/kata-containers/pull/1389)) so the port may require some rework.

fidencio commented:
Folks, this PR has been stale and I'm closing it as we're walking towards the final 1.13.0 release of the kata-containers-1.x.
From now on, all the development will happen as part of https://github.com/kata-containers/kata-containers/ and fixes should be provided to that repo.

Thanks a lot for your contribution, we're sorry that for some reason it got stale, and we hope to see you around for new adventures in the kata-containers-2.x land.
